docs/designs/keras_callback.md

```diff
@@ -0,0 +1,196 @@
+# Design to Support Keras Callback
+
+This document describes the design for supporting callback to customize the behavior of model in ElasticDL.
+
+## Motivation
+
+In deep learning, we generally need to customize the behavior of model during training, evaluation or
+inference, including reading/changing the model. According to the call frequency, we may perform the behavior per batch, per epoch or at the start and end of job. `tf.keras.callbacks.Callback` is an abstract base class and has methods to perform the behavior at different call frequency, such as `on_bath_end`, `on_epoch_end` and so on. So, we adopt the interfaces of `tf.keras.callbacks.Callback` for users to customize the behavior of model in ElasticDL. 
+
+Now we have implemented some modules which are similar to callback in ElasticDL such as `LearingRateScheduler` and `PredictionOutputsProcessor`. And user should write definitions
+in the model definition file for each module like:
+```python
+def custom_model():
+    ...
+
+# Adjust the learning rate according to iteration steps
+def learning_rate_scheduler(model_version):
+    if model_version < 5000:
+        return 0.0003
+    elif model_version < 12000:
+        return 0.0002
+    else:
+        return 0.0001
+
+# Process the prediction outputs for each batch
+class PredictionOutputsProcessor(BasePredictionOutputsProcessor):
+    ...
+```
+There will be multiple definition APIs for users to define different behaviors of model and the interfaces of APIs may be different. It is more convenient for users to define those behaviors using `tf.keras.callbacks.Callback`. 
+
+Some use cases we observed that we want to support are:
+* Callbacks used in testing that checks retry times and model parameter updates,
+* Callback similar to `PredictionOutputsProcessor` that is executed after prediction outputs are made.
+* Callback to write additional summaries to TensorBoard after each evaluation job completes.
+* Callback to modulate learning rate as we are currently doing in `LearningRateScheduler`.
+* Callback for logic to be executed before training starts such as starting embedding service.
+* Callback to perform early stopping when certain conditions are met.
+* Callback to upload model to remote storage after completing training.
+
```

Comment:
How about the current checkpointing strategy? Can they be implemented as callback as well?

Reply:
Yes, the checkpointing strategy can also be implemented as a callback which can call a method on the PS every updating gradients of a batch. Maybe we need `on_update_batch_end` to implement checkpointing strategy.

Reply:
Due to we are developing PS using Golang, the current checkpointing strategy can not be implemented as callback using ParameterServerStrategy.